### PR TITLE
 add setLogDelegateFactory() to LoggerFactory, fixes #2298

### LINKED ISSUE #2298: Make LogDelegateFactory configurable via VertxOptions to avoid classloading issues in ContextClassLoader sensitive environments such as OSGi
The use of:

```java
String className = System.getProperty("vertx.logger-delegate-factory-class-name");
Thread.currentThread().getContextClassLoader().loadClass(className);
```

in `io.vertx.core.logging.LoggerFactory` can be problematic in "context ClassLoader sensitive environments", such as OSGi (but this could also affect use of Vert.x within other environments that do funky class loading); here are some links with background about why that is so:

* http://njbartlett.name/2010/08/30/osgi-readiness-loading-classes.html
* http://wiki.eclipse.org/Context_Class_Loader_Enhancements
* https://helpx.adobe.com/experience-manager/kb/OsgiClassLoading3Party.html
* https://issues.apache.org/jira/browse/FELIX-2951

I've actually run into this problem myself, see the detailed error below (which probably won't really make all that sense, without more context ; it's from https://github.com/vorburger/minecraft-storeys-maker/ running under https://github.com/vorburger/ch.vorburger.minecraft.osgi ... but ignore that).

I was able to work-around this in my code by instead of doing this:

```java
System.setProperty("vertx.logger-delegate-factory-class-name", SLF4JLogDelegateFactory.class.getName());
vertx = Vertx.vertx(new VertxOptions().setWorkerPoolSize(23));
```

using this beauty:

```java
ClassLoader tccl = Thread.currentThread().getContextClassLoader();
Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
try {
    System.setProperty("vertx.logger-delegate-factory-class-name", SLF4JLogDelegateFactory.class.getName());
    vertx = Vertx.vertx(new VertxOptions().setWorkerPoolSize(23));
} finally {
    Thread.currentThread().setContextClassLoader(tccl);
}
```

but in an ideal world... this kind of hand-stand should not be required?

One possible solution to this could be to offer an alternative to the system property and ClassLoader mess by simply letting users of Vert.x specify an instance of their desired `io.vertx.core.spi.logging.LogDelegateFactory` programmatically - I guess this would naturally fit into the `io.vertx.core.VertxOptions.VertxOptions` ?

Another solution, less flexible and IMHO best combined with above, would be to let `io.vertx.core.logging.LoggerFactory` try both (first) the `getContextClassLoader()` (as well as, if NOK) the `LoggerFactory.class.getClassLoader()`.

Would a Pull Request proposing this via code be a welcome contribution to the Vert.x project?

**Full background:**

```
[00:07:57 ERROR] [ch.vorburger.osgi.builder.internal.SourceInstallServiceImpl]: Problem reading/installing bundle JAR: /home/vorburger/dev/Minecraft/git/minecraft-storeys-maker/web/build/libs/web-1.0.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Activator start error in bundle web [30].
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2276) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2144) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at ch.vorburger.osgi.builder.internal.SourceInstallServiceImpl.lambda$installSourceBundle$0(SourceInstallServiceImpl.java:69) ~[4_ch.vorburger.osgi.gradle-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.osgi.builder.internal.SourceInstallServiceImpl.lambda$installSourceBundle$1(SourceInstallServiceImpl.java:102) ~[4_ch.vorburger.osgi.gradle-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.fswatch.DirectoryWatcherBuilder.firstListenerNotification(DirectoryWatcherBuilder.java:86) [4_ch.vorburger.osgi.gradle-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.fswatch.FileWatcherBuilder.build(FileWatcherBuilder.java:53) [4_ch.vorburger.osgi.gradle-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.osgi.builder.internal.SourceInstallServiceImpl.installSourceBundle(SourceInstallServiceImpl.java:117) [4_ch.vorburger.osgi.gradle-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.minecraft.osgi.dev.BundleManagerPersistence.installAndStartAll(BundleManagerPersistence.java:71) [5_ch.vorburger.minecraft.osgi.dev-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.minecraft.osgi.dev.BundleManager.<init>(BundleManager.java:50) [5_ch.vorburger.minecraft.osgi.dev-1.0.0-SNAPSHOT.jar:?]
	at ch.vorburger.minecraft.osgi.dev.internal.Activator.start(Activator.java:38) [5_ch.vorburger.minecraft.osgi.dev-1.0.0-SNAPSHOT.jar:?]
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) [ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2226) [ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2144) [ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) [ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) [ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at ch.vorburger.minecraft.osgi.OSGiFrameworkWrapper.installBundles(OSGiFrameworkWrapper.java:158) [OSGiFrameworkWrapper.class:?]
	at ch.vorburger.minecraft.osgi.OSGiFrameworkWrapper.installBundles(OSGiFrameworkWrapper.java:134) [OSGiFrameworkWrapper.class:?]
	at ch.vorburger.minecraft.osgi.OSGiFrameworkWrapper.installBootBundles(OSGiFrameworkWrapper.java:123) [OSGiFrameworkWrapper.class:?]
	at ch.vorburger.minecraft.osgi.Bootstrap.bootstrapMinecraftOSGi(Bootstrap.java:50) [Bootstrap.class:?]
	at ch.vorburger.minecraft.osgi.MinecraftSpongePlugin.onGameStartingServerEvent(MinecraftSpongePlugin.java:52) [MinecraftSpongePlugin.class:?]
	at org.spongepowered.common.event.listener.GameStartingServerEventListener_MinecraftSpongePlugin_onGameStartingServerEvent3.handle(Unknown Source) [?:?]
	at org.spongepowered.common.event.RegisteredListener.handle(RegisteredListener.java:95) [RegisteredListener.class:1.12.2-7.1.0-BETA-13]
	at org.spongepowered.common.event.SpongeEventManager.post(SpongeEventManager.java:403) [SpongeEventManager.class:1.12.2-7.1.0-BETA-13]
	at org.spongepowered.common.event.SpongeEventManager.post(SpongeEventManager.java:430) [SpongeEventManager.class:1.12.2-7.1.0-BETA-13]
	at org.spongepowered.common.SpongeImpl.postEvent(SpongeImpl.java:213) [SpongeImpl.class:1.12.2-7.1.0-BETA-13]
	at org.spongepowered.common.SpongeImpl.postState(SpongeImpl.java:221) [SpongeImpl.class:1.12.2-7.1.0-BETA-13]
	at org.spongepowered.server.SpongeVanilla.onServerStarting(SpongeVanilla.java:162) [SpongeVanilla.class:1.12.2-7.1.0-BETA-13]
	at net.minecraft.server.dedicated.DedicatedServer.handler$callServerStarting$zpk000(SourceFile:1258) [nz.class:?]
	at net.minecraft.server.dedicated.DedicatedServer.func_71197_b(SourceFile:233) [nz.class:?]
	at net.minecraft.server.MinecraftServer.run(SourceFile:434) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
Caused by: java.lang.ExceptionInInitializerError
	at io.vertx.core.impl.VertxImpl.<clinit>(VertxImpl.java:99) ~[?:?]
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:42) ~[?:?]
	at io.vertx.core.Vertx.vertx(Vertx.java:92) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.VertxStarter.start(VertxStarter.java:49) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.StoreysWebPlugin.start(StoreysWebPlugin.java:66) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.Activator.start(Activator.java:44) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2226) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	... 32 more
Caused by: java.lang.IllegalArgumentException: Error instantiating transformer class "io.vertx.core.logging.SLF4JLogDelegateFactory"
	at io.vertx.core.logging.LoggerFactory.initialise(LoggerFactory.java:59) ~[?:?]
	at io.vertx.core.logging.LoggerFactory.<clinit>(LoggerFactory.java:37) ~[?:?]
	at io.vertx.core.impl.VertxImpl.<clinit>(VertxImpl.java:99) ~[?:?]
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:42) ~[?:?]
	at io.vertx.core.Vertx.vertx(Vertx.java:92) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.VertxStarter.start(VertxStarter.java:49) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.StoreysWebPlugin.start(StoreysWebPlugin.java:66) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.Activator.start(Activator.java:44) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2226) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	... 32 more
Caused by: java.lang.ClassNotFoundException: io.vertx.core.logging.SLF4JLogDelegateFactory
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_151]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_151]
	at io.vertx.core.logging.LoggerFactory.initialise(LoggerFactory.java:56) ~[?:?]
	at io.vertx.core.logging.LoggerFactory.<clinit>(LoggerFactory.java:37) ~[?:?]
	at io.vertx.core.impl.VertxImpl.<clinit>(VertxImpl.java:99) ~[?:?]
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:42) ~[?:?]
	at io.vertx.core.Vertx.vertx(Vertx.java:92) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.VertxStarter.start(VertxStarter.java:49) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.StoreysWebPlugin.start(StoreysWebPlugin.java:66) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.Activator.start(Activator.java:44) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2226) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	... 32 more
Caused by: java.lang.NullPointerException
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:182) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_151]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_151]
	at io.vertx.core.logging.LoggerFactory.initialise(LoggerFactory.java:56) ~[?:?]
	at io.vertx.core.logging.LoggerFactory.<clinit>(LoggerFactory.java:37) ~[?:?]
	at io.vertx.core.impl.VertxImpl.<clinit>(VertxImpl.java:99) ~[?:?]
	at io.vertx.core.impl.VertxFactoryImpl.vertx(VertxFactoryImpl.java:42) ~[?:?]
	at io.vertx.core.Vertx.vertx(Vertx.java:92) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.VertxStarter.start(VertxStarter.java:49) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.StoreysWebPlugin.start(StoreysWebPlugin.java:66) ~[?:?]
	at ch.vorburger.minecraft.storeys.web.Activator.start(Activator.java:44) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2226) ~[ch.vorburger.minecraft.osgi-1.0.0-SNAPSHOT-all.jar:?]
	... 32 more
```

### PR DESCRIPTION
**Pull request by vorburger**

### CODE REVIEW

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 1]
```diff
@@ -46,18 +46,55 @@ public static synchronized void initialise() {
...
-    LoggerFactory.delegateFactory = delegateFactory;
+    LoggerFactory.delegateFactory = newDelegateFactory;
+  }
+
+  private static LogDelegateFactory loadClass(String className, ClassLoader classLoader) {
+    try {
+      Class<?> clz = classLoader.loadClass(className);
+      return (LogDelegateFactory) clz.newInstance();
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Error instantiating transformer class \"" + className + "\"", e);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Why _transformer_?

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 2]
```diff
@@ -46,18 +46,55 @@ public static synchronized void initialise() {
...
+    }
+  }
+
+  /**
+   * Set new custom LoggerFactory.
+   * @param factory the new LogDelegateFactory to use
+   * @author Michael Vorburger.ch
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Remove the `@Author` tag from method javadoc.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 3]
```diff
@@ -46,18 +46,55 @@ public static synchronized void initialise() {
...
+   */
+  public static void setLogDelegateFactory(LogDelegateFactory factory) {
+      LogDelegateFactory current = delegateFactory;
+      if (current != null) {
+          try {
+            getLogger(LoggerFactory.class).info("Replacing LogDelegateFactory " + current + " by " + factory);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Not sure of `info`, this looks like a `debug` message

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 4]
```diff
@@ -46,18 +46,55 @@ public static synchronized void initialise() {
...
+    }
+  }
+
+  /**
+   * Set new custom LoggerFactory.
+   * @param factory the new LogDelegateFactory to use
+   * @author Michael Vorburger.ch
+   */
+  public static void setLogDelegateFactory(LogDelegateFactory factory) {
+      LogDelegateFactory current = delegateFactory;
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: It should be synchronized.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 5]
```diff
@@ -46,18 +46,55 @@ public static synchronized void initialise() {
...
+      if (current != null) {
+          try {
+            getLogger(LoggerFactory.class).info("Replacing LogDelegateFactory " + current + " by " + factory);
+          } catch (Throwable t) {
+            // Too bad...
+            t.printStackTrace();
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Can't we use a logger of `factory` to manage this stack trace?

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 4 above)
Comment (cescoffier) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: Shouldn't we test that factory is not null.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 1 above)
Comment (vorburger) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: dunno, but did you noticed I've just moved this (from line 54), did not actually write it... but yeah, it does looks weird, probably an old copy/paste mistake; so I've just changed it as part of this commit.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 2 above)
Comment (vorburger) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: sure, done

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 4 above)
Comment (vorburger) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: added `synchronized`, and null checked `factory` argument

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 3 above)
Comment (vorburger) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: sure, replaced `info` by `debug`

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 6]
```diff
@@ -69,6 +69,34 @@ private static LogDelegateFactory loadClass(String className, ClassLoader classL
...
+          }
+          loggers.clear();
+      }
+      delegateFactory = factory;
+  }
+
+  /**
+   * Get current LoggerFactory.
+   * @return the currently used LogDelegateFactory
+   * @author Michael Vorburger.ch
```
Comment (vorburger) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: FYI I've removed `@author` here as well..

File: src/main/java/io/vertx/core/logging/LoggerFactory.java (hunk 5 above)
Comment (vorburger) [reply src/main/java/io/vertx/core/logging/LoggerFactory.java]: duh, yeah ... nice idea - I simply didn't think of it; now done as suggested

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 7]
```diff
@@ -46,18 +47,54 @@ public static synchronized void initialise() {
...
+   * @param factory the new LogDelegateFactory to use
+   */
+  public static synchronized void setLogDelegateFactory(LogDelegateFactory factory) {
+      requireNonNull(factory, "factory == null");
+      LogDelegateFactory current = delegateFactory;
+      if (current != null) {
+          try {
+            getLogger(LoggerFactory.class).debug("Replacing LogDelegateFactory " + current + " by " + factory);
+          } catch (Throwable t) {
+            factory.createDelegate(LoggerFactory.class.getCanonicalName())
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Why are you using the canonical name here? We generally use just the FQCN.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 8]
```diff
@@ -46,18 +47,54 @@ public static synchronized void initialise() {
...
+
+    LoggerFactory.delegateFactory = newDelegateFactory;
+  }
+
+  private static LogDelegateFactory loadClass(String className, ClassLoader classLoader) {
+    try {
+      Class<?> clz = classLoader.loadClass(className);
+      return (LogDelegateFactory) clz.newInstance();
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Error instantiating LogDelegateFactory implemention: \"" + className + "\"", e);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: There is a few issues here:

* The class cannot be loaded
* The access is not allowed 
* The constructor throws an exception

We may want to throw different type of exceptions for these types.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 9]
```diff
@@ -46,18 +47,54 @@ public static synchronized void initialise() {
...
-      ClassLoader loader = Thread.currentThread().getContextClassLoader();
+      ClassLoader tcccl = Thread.currentThread().getContextClassLoader();
       try {
-        Class<?> clz = loader.loadClass(className);
-        delegateFactory = (LogDelegateFactory) clz.newInstance();
-      } catch (Exception e) {
-        throw new IllegalArgumentException("Error instantiating transformer class \"" + className + "\"", e);
+        newDelegateFactory = loadClass(className, tcccl);
+      } catch (IllegalArgumentException e) {
+        ClassLoader thisClasssLoader = LoggerFactory.class.getClassLoader();
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: Should we log something here (maybe debug) indicating that the initial loading failed? The issue is that if it's a ClassNotFound it's fine. But in the 2 other cases, it should be a showstopper.

File: src/main/java/io/vertx/core/logging/LoggerFactory.java [hunk 10]
```diff
@@ -46,18 +47,54 @@ public static synchronized void initialise() {
...
+      }
+      delegateFactory = factory;
+  }
 
-    LoggerFactory.delegateFactory = delegateFactory;
+  /**
+   * Get current LoggerFactory.
+   * @return the currently used LogDelegateFactory
+   */
+  public static LogDelegateFactory getLogDelegateFactory() {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/logging/LoggerFactory.java]: need to be synchronized

**Comment by vorburger** [comment]
Hello @cescoffier I'm doing a year end clean up of my personal https://github.com/pulls and wanted to ask if you would still consider merging this very old Pull Request if I rebased it, or if we should forget about and close this? (This is a bulk message I'm posting to many issues to gauge what is still revelant.)

**Pull request closed without merging**